#! /bin/sh -e
 
# DP: biarch patches for ppc64
 
dir=
if [ $# -eq 3 -a "$2" = '-d' ]; then
    pdir="-d $3"
    dir="$3/"
elif [ $# -ne 1 ]; then
    echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
    exit 1
fi
case "$1" in
    -patch)
        patch $pdir -f --no-backup-if-mismatch -p1 < $0
        ;;
    -unpatch)
        patch $pdir -f --no-backup-if-mismatch -R -p1 < $0
        ;;
    *)
        echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
        exit 1
esac
exit 0


diff -urN tmp/gcc/config/rs6000/t-linux64 src/gcc/config/rs6000/t-linux64
--- tmp/gcc/config/rs6000/t-linux64	2004-12-04 08:05:46.000000000 +0000
+++ src/gcc/config/rs6000/t-linux64	2004-12-04 08:07:52.180889736 +0000
@@ -8,13 +8,10 @@
 
 SHLIB_MAPFILES += $(srcdir)/config/rs6000/libgcc-ppc64.ver
 
-MULTILIB_OPTIONS        = m64/m32 msoft-float
-MULTILIB_DIRNAMES       = 64 32 nof
+MULTILIB_OPTIONS        = m64/m32
+MULTILIB_DIRNAMES       = 64 32
 MULTILIB_EXTRA_OPTS     = fPIC mstrict-align
-MULTILIB_EXCEPTIONS     = m64/msoft-float
-MULTILIB_EXCLUSIONS     = m64/!m32/msoft-float
-MULTILIB_OSDIRNAMES	= ../lib64 ../lib nof
+MULTILIB_OSDIRNAMES	= ../lib ../lib32
-MULTILIB_MATCHES        = $(MULTILIB_MATCHES_FLOAT)
 
 # We want fine grained libraries, so use the new code to build the
 # floating point emulation libraries.
@@ -37,8 +34,11 @@
 mklibgcc: bispecs
 
 bispecs: specs
-	if [ x`$(GCC_FOR_TARGET) -print-multi-os-directory` = x../lib ]; then \
+	touch f-test.c; \
+	$(GCC_FOR_TARGET) -c f-test.c -o f-test.o; \
+	if [ "x`file f-test.o | grep 64-bit`" = "x" ]; then \
 	  sed -e '/cc1_options/{ n; s/$$/ %{m64:-mlong-double-128}/; }' < specs > $@; \
 	else \
 	  sed -e '/cc1_options/{ n; s/$$/ %{!m32:-mlong-double-128}/; }' < specs > $@; \
-	fi
+	fi; \
+	rm f-test.c f-test.o;
diff -urN tmp/libjava/configure.host src/libjava/configure.host
--- tmp/libjava/configure.host	25 Nov 2004 03:46:56 -0000
+++ src/libjava/configure.host	15 Dec 2004 15:45:22 -0000
@@ -123,10 +123,13 @@
   powerpc64*-*)
 	sysdeps_dir=powerpc
 	libgcj_interpreter=yes
-	if [ x`$CC -print-multi-os-directory` = x../lib64 ]; then
+	touch f-test.c
+	$CC -c f-test.c -o f-test.o
+	if [ "x`file f-test.o | grep 32-bit`" = "x" ]; then
 	    libgcj_flags="${libgcj_flags} -mminimal-toc"
	fi
 	enable_hash_synchronization_default=yes
 	slow_pthread_self=yes
+	rm f-test.c f-test.o
 	;;
   powerpc*-*)
diff -urN tmp/config-ml.in src/config-ml.in
--- tmp/config-ml.in	2004-11-09 08:07:29.000000000 +0000
+++ src/config-ml.in	2004-12-12 07:12:44.631147816 +0000
@@ -400,7 +400,47 @@
 	  ;;
 	esac
 	;;
-powerpc*-*-* | rs6000*-*-*)
+powerpc64-*-*)
+ 	case " $multidirs " in
+ 	*" 32 "*)
+ 	  # We will not be able to create libraries with -m32 if
+ 	  # we cannot even link a trivial program.  It usually
+ 	  # indicates the 32bit libraries are missing.
+ 	  if echo 'main() {}' > conftest.c &&
+ 	     ${CC-gcc} -m32 conftest.c -o conftest; then
+ 	   echo Enable only libstdc++.
+ 	   old_multidirs="${multidirs}"
+ 	   multidirs=""
+ 	   for x in ${old_multidirs}; do
+ 	     case "$x" in
+ 	     *32* ) case ${ml_realsrcdir} in
+ 	               *"libstdc++-v3" ) multidirs="${multidirs} ${x}" ;;
+ 	               *"libgfortran" ) multidirs="${multidirs} ${x}" ;;
+ 	               *"libobjc" ) multidirs="${multidirs} ${x}" ;;
+ 	               *"libiberty" ) multidirs="${multidirs} ${x}" ;;
+ 	               *"zlib" ) multidirs="${multidirs} ${x}" ;;
+ 	               *) : ;;
+ 	            esac 
+ 	            ;;
+ 	     *) multidirs="${multidirs} ${x}" ;;
+ 	     esac
+ 	   done
+ 	  else
+ 	    echo Could not link program with -m32, disabling it.
+ 	    old_multidirs="${multidirs}"
+ 	    multidirs=""
+ 	    for x in ${old_multidirs}; do
+ 	      case "$x" in
+ 	      *m32* ) : ;;
+ 	      *) multidirs="${multidirs} ${x}" ;;
+ 	      esac
+ 	    done
+ 	  fi
+ 	  rm -f conftest.c conftest
+ 	  ;;
+ 	esac
+ 	;;
+powerpc-*-* | rs6000*-*-*)
 	if [ x$enable_aix64 = xno ]
 	then
 	  old_multidirs="${multidirs}"
